README.md
# Map Controller Plus v3.0.0

Stateful map controller for Flutter Map v3.0.0. Manage markers, lines and polygons.

**This is a fork from [synw's map_controller package](https://pub.dev/packages/map_controller_plus) made because the project has been abandoned. This new and improved version supports the latest version of the [flutter_map](https://pub.dev/packages/flutter_map) package. If you need a feature or a fix you can [open an issue on the forked repository](https://github.com/makesappy/map_controller_plus_3/issues).**

## Usage

```dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:latlong/latlong.dart';
import 'package:map_controller/map_controller.dart';

class MapPage extends StatefulWidget {
   @override
   State<MapPage> createState() => _MapPageState();
}

class _MapPageState extends State<MapPage> {
   late final MapController mapController;
   late final StatefulMapController statefulMapController;
   late final StreamSubscription<StatefulMapControllerStateChange> sub;
   
   @override
   void initState() {
      super.initState();

      // intialize the controllers
      mapController = MapController();
      statefulMapController = StatefulMapController(mapController: mapController);

      // wait for the controller to be ready before using it
      statefulMapController.onReady.then((_) => print("The map controller is ready")));

      /// [Important] listen to the changefeed to rebuild the map on changes:
      /// this will rebuild the map when for example addMarker or any method 
      /// that mutates the map assets is called
      sub = statefulMapController.changeFeed.listen((change) => setState(() {}));
   }
   
   @override
   Widget build(BuildContext context) {
      return Scaffold(
         body: SafeArea(
            child: Stack(children: <Widget>[
               FlutterMap(
                  mapController: mapController,
                  options: MapOptions(center: LatLng(48.853831, 2.348722), zoom: 11.0),
                  layers: [
                     MarkerLayer(markers: statefulMapController.markers),
                     PolylineLayer(polylines: statefulMapController.lines),
                     PolygonLayer(polygons: statefulMapController.polygons),
                  ],
               ),
            // ...
         ])),
      );
   }
   
   @override
   void dispose() {
      sub.cancel();
      super.dispose();
   }
}
```

## Api

Api for the [StatefulMapController](https://pub.dev/documentation/map_controller/latest/map_controller/StatefulMapController-class.html) class

### Map controls

#### Zoom

* `zoom`: get the current zoom value
* `zoomIn()`: increase the zoom level by 1
* `zoomOut()`: decrease the zoom level by 1
* `zoomTo()`: zoom to the provided value

#### Center

* `center`: get the current center `LatLng` value
* `centerOnPoint()`: center on the `LatLng` value

### Map assets

#### Markers

* `addMarker()`: add a named marker on the map
* `addMarkers()`: add several named markers on the map
* `removeMarker()`: remove a named marker from the map
* `removeMarkers()`: remove several named markers from the map
* `markers`: get the markers that are on the map
* `namedMarkers`: get the markers with their names that are on the map
* `getMarker()`: return the marker with the corresponding name
* `getMarkers()`: return the markers with the corresponding names

#### Stateful markers

*New in 0.7*: the stateful makers hold their own state and can be mutated 

```dart
statefulMapController.addStatefulMarker(
   name: "some marker",
   statefulMarker: StatefulMarker(
      height: 80.0,
      width: 120.0,
      state: <String, dynamic>{"showText": false},
      point: LatLng(48.853831, 2.348722),
      builder: (BuildContext context, Map<String, dynamic> state) {
         Widget w;
         final markerIcon = IconButton(
            icon: const Icon(Icons.location_on),
            onPressed: () => statefulMapController.mutateMarker(
               name: "some marker",
               property: "showText",
               value: !(state["showText"] as bool)));
         if (state["showText"] == true) {
            w = Column(children: <Widget>[
               markerIcon,
               Container(
                  color: Colors.white,
                  child: Padding(
                     padding: const EdgeInsets.all(5.0),
                     child: Text(place.name, textScaleFactor: 1.3))),
            ]);
         } else {
            w = markerIcon;
         }
         return w;
      })
);
```

#### Lines

* `addLine()`: add a line on the map
* `lines`: get the lines that are on the map

#### Polygons

* `addPolygon`: add a polygon on the map
* `polygons`: get the polygons that are on the map

### On ready callback

Execute some code right after the map is ready:

```dart
@override
void initState() {
   super.initState();
   statefulMapController.onReady.then((_) {
      setState((_) =>_ready = true);
   });
}
```

### Changefeed

A changefeed is available: it's a stream with all state changes from the map controller. Use it to update the map when a change occurs:

```dart
statefulMapController.onReady.then((_) {
   statefulMapController.changeFeed.listen((change) => setState(() {}));
});
```

### Geojson data

The map controller can draw on the map from geojson data:

```dart
void loadData() async {
  print("Loading geojson data");
  final data = await rootBundle.loadString('assets/airports.geojson');
  await statefulMapController.fromGeoJson(data,
    markerIcon: Icon(Icons.local_airport), verbose: true);
}

@override
void initState() {
  mapController = MapController();
  statefulMapController = StatefulMapController(mapController: mapController);
  statefulMapController.onReady.then((_) => loadData());
  sub = statefulMapController.changeFeed.listen((change) => setState(() {}));
  super.initState();
}
```

### Tile layer management

Some predefined tile layers are available.

```dart
FlutterMap(
   mapController: mapController,
   options: MapOptions(
      center: LatLng(48.853831, 2.348722),
      zoom: 11.0,
   ),
   layers: [
      // Use the map controller's tile layer
      statefulMapController.tileLayer,
      MarkerLayer(markers: statefulMapController.markers),
      
      // ...
   ],
)
```

To switch tile layers at runtime use:

```dart
statefulMapController.switchTileLayer(TileLayerType.monochrome);
```

Available layers:

```dart
TileLayerType.normal
TileLayerType.topography
TileLayerType.monochrome
TileLayerType.hike
```

A tile layers bar is available:

```dart
Stack(
   children: <Widget>[
      FlutterMap(
         mapController: mapController,
         options: MapOptions(
            center: LatLng(48.853831, 2.348722),
            zoom: 11.0,
         ),
         layers: [
            statefulMapController.tileLayer,
            MarkerLayer(markers: statefulMapController.markers),
         ],
      ),
      Positioned(
         top: 15.0,
         right: 15.0,
         child: TileLayersBar(controller: statefulMapController),
      ),
   ],
);
```


CHANGELOG.md
# Changelog

## 1.0.0

* Fork of the [synw's map_controller package](https://pub.dev/packages/map_controller_plus) lib
* flutter_map upgraded to version 3.0.0 with all needed changes and migrations


test/utils/test_app.dart
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:latlong2/latlong.dart';

class TestApp extends StatefulWidget {
  final MapController? mapController;

  /// Test widget to show a map with a marker with a zoom level of `13.0`.
  const TestApp({Key? key, this.mapController}) : super(key: key);

  @override
  State<TestApp> createState() => _TestAppState();
}

class _TestAppState extends State<TestApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: FlutterMap(
          mapController: widget.mapController,
          options: MapOptions(
            center: LatLng(45.5231, -122.6765),
            zoom: 13.0,
          ),
          children: [
            TileLayer(
              urlTemplate: 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
              subdomains: ['a', 'b', 'c'],
            ),
            MarkerLayer(
              markers: [
                Marker(
                  width: 80.0,
                  height: 80.0,
                  point: LatLng(45.5231, -122.6765),
                  builder: (ctx) => const FlutterLogo(),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}


test/utils/custom_tests.dart
import 'dart:io';

import 'package:flutter_map/flutter_map.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:map_controller_plus_3/src/controller.dart';
import 'package:meta/meta.dart';

import 'mock_http_client.dart';

typedef FlutterMapTesterCallback = Future<void> Function(
  WidgetTester widgetTester,
  MapController mapController,
  StatefulMapController statefulMapController,
);

/// Custom test to expose a [MapController] and [StatefulMapController], it also
/// overrides the global [HttpOverrides].
@isTest
void testFlutterMap(String description, FlutterMapTesterCallback callback) {
  testWidgets(description, (tester) async {
    final mapController = MapController();
    final statefulController = StatefulMapController(
      mapController: mapController,
    );
    HttpOverrides.global = MockHttpOverrides();
    await callback(tester, mapController, statefulController);
  });
}


lib/src/controller.dart
import 'dart:async';

import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:geojson/geojson.dart';
import 'package:geopoint/geopoint.dart';
import 'package:latlong2/latlong.dart';
import 'package:rxdart/rxdart.dart';

import '../map_controller_plus_3.dart';
import 'exceptions.dart';
import 'state/lines.dart';
import 'state/map.dart';
import 'state/markers.dart';
import 'state/polygons.dart';
import 'state/stateful_markers.dart';
import 'state/tile_layer.dart';

/// Function to notify the changefeed
typedef FeedNotifyFunction = void Function(
  String name,
  Object? value,
  Function from,
  MapControllerChangeType type,
);

/// The map controller
class StatefulMapController {
  /// Provide a Flutter map [MapController]
  StatefulMapController({
    required this.mapController,
    this.tileLayerType = TileLayerType.normal,
    this.customTileLayer,
    this.verbose = false,
  }) {
    // init state
    _markersState = MarkersState(mapController: mapController, notify: notify);
    _linesState = LinesState(notify: notify);
    _polygonsState = PolygonsState(notify: notify);
    _mapState = MapState(mapController: mapController, notify: notify);
    _statefulMarkersState =
        StatefulMarkersState(mapController: mapController, notify: notify);
    if (customTileLayer != null) {
      tileLayerType = TileLayerType.custom;
    }
    _tileLayerState = TileLayerState(
      type: tileLayerType,
      customTileLayer: customTileLayer,
      notify: notify,
    );
    WidgetsBinding.instance.addPostFrameCallback((_) {
      // fire the map is ready callback
      if (!_readyCompleter.isCompleted) {
        _readyCompleter.complete();
      }
    });
  }

  /// The Flutter Map [MapController]
  final MapController mapController;

  /// The Flutter Map [MapOptions]
  MapOptions? mapOptions;

  /// The initial tile layer
  TileLayerType tileLayerType;

  /// A custom tile layer options
  TileLayer? customTileLayer;

  /// Verbosity level
  final bool verbose;

  late MapState _mapState;
  late MarkersState _markersState;
  late LinesState _linesState;
  late PolygonsState _polygonsState;
  late TileLayerState _tileLayerState;
  late StatefulMarkersState _statefulMarkersState;

  final _readyCompleter = Completer<void>();
  final _subject = PublishSubject<StatefulMapControllerStateChange>();

  /// On ready callback: this is fired when the contoller is ready
  Future<void> get onReady => _readyCompleter.future;

  /// A stream with changes occuring on the map
  Stream<StatefulMapControllerStateChange> get changeFeed =>
      _subject.distinct();

  /// The map zoom value
  double get zoom => mapController.zoom;

  /// Rotate the map
  set rotate(double degree) => mapController.rotate(degree);

  /// The map center value
  LatLng get center => mapController.center;

  /// The stateful markers present on the map
  Map<String, StatefulMarker> get statefulMarkers =>
      _statefulMarkersState.statefulMarkers;

  void addStatefulMarker({
    required String name,
    required StatefulMarker statefulMarker,
  }) =>
      _statefulMarkersState.addStatefulMarker(name, statefulMarker);

  void addStatefulMarkers(Map<String, StatefulMarker> statefulMarkers) =>
      _statefulMarkersState.addStatefulMarkers(statefulMarkers);

  void mutateMarker({
    required String name,
    required String property,
    required Object? value,
  }) =>
      _statefulMarkersState.mutate(name, property, value);

  /// The markers present on the map
  List<Marker> get markers {
    final localMarkers = <Marker>[];
    localMarkers
      ..addAll(_markersState.markers)
      ..addAll(_statefulMarkersState.markers);
    return localMarkers;
  }

  /// Return a [Marker] corresponding to [name] from the [StatefulMarkersState].
  Marker? getMarker(String name) {
    final marker = _statefulMarkersState.statefulMarkers[name];
    return marker?.marker;
  }

  /// Return all [Marker] which have their name in [names].
  ///
  /// If one of the name doesn't correspond to any marker it is not added
  /// to the returned list.
  ///
  /// If no markers were found return an empty list.
  List<Marker> getMarkers(List<String> names) {
    final markers = <Marker>[];
    final localStatefulMarkers = _statefulMarkersState.statefulMarkers;
    for (final name in names) {
      final marker = localStatefulMarkers[name];
      if (marker != null) markers.add(marker.marker);
    }
    return markers;
  }

  /// The markers present on the map and their names
  Map<String, Marker> get namedMarkers => _markersState.namedMarkers;

  /// The lines present on the map
  List<Polyline> get lines => _linesState.lines;

  /// The named lines present on the map
  Map<String, Polyline> get namedLines => _linesState.namedLines;

  /// Return a [Polyline] corresponding to [name] from the [LinesState].
  ///
  /// If no corresponding line was found it will return [null].
  Polyline? getLine(String name) => _linesState.namedLines[name];

  /// Return all [Polyline] which have their name in [names].
  ///
  /// If one of the name doesn't correspond to any marker it is not added
  /// to the returned list.
  ///
  /// If no markers were found return an empty list.
  List<Polyline> getLines(List<String> names) {
    final lines = <Polyline>[];
    for (final name in names) {
      final line = getLine(name);
      if (line != null) lines.add(line);
    }
    return lines;
  }

  /// The polygons present on the map
  List<Polygon> get polygons => _polygonsState.polygons;

  /// The named polygons present on the map
  Map<String, Polygon> get namedPolygons => _polygonsState.namedPolygons;

  /// The current map tile layer
  TileLayer? get tileLayer => _tileLayerState.tileLayer;

  /// Zoom in one level
  void zoomIn() => _mapState.zoomIn();

  /// Zoom out one level
  void zoomOut() => _mapState.zoomOut();

  /// Zoom to level
  void zoomTo(double value) => _mapState.zoomTo(value);

  /// Center the map on a [LatLng]
  void centerOnPoint(LatLng point) => _mapState.centerOnPoint(point);

  /// The callback used to handle gestures and keep the state in sync
  void onPositionChanged(MapPosition pos, {required bool gesture}) =>
      _mapState.onPositionChanged(pos, gesture: gesture);

  /// Add a marker on the map
  void addMarker({required Marker marker, required String name}) =>
      _markersState.addMarker(marker: marker, name: name);

  /// Remove a marker from the map
  void removeMarker({required String name}) =>
      _markersState.removeMarker(name: name);

  /// Add multiple markers to the map
  void addMarkers({required Map<String, Marker> markers}) =>
      _markersState.addMarkers(markers: markers);

  /// Remove multiple makers from the map
  void removeMarkers({required List<String> names}) =>
      _markersState.removeMarkers(names: names);

  /// Fit bounds for all markers on map
  Future<void> fitMarkers() async => _markersState.fitAll();

  /// Fit bounds for one marker on map
  Future<void> fitMarker(String name) async => _markersState.fitOne(name);

  /// Fit bounds and zoom the map to center on a line
  void fitLine(String name) {
    final line = _linesState.namedLines[name]!;
    final bounds = LatLngBounds();
    for (final point in line.points) {
      bounds.extend(point);
    }
    mapController.fitBounds(bounds);
  }

  /// Add a line on the map
  void addLine({
    required String name,
    required List<LatLng> points,
    double width = 3.0,
    Color color = Colors.green,
    bool isDotted = false,
  }) {
    _linesState.addLine(
      name: name,
      line: Polyline(
        points: points,
        color: color,
        strokeWidth: width,
        isDotted: isDotted,
      ),
    );
  }

  /// Add a line on the map
  void addLineFromGeoPoints({
    required String name,
    required List<GeoPoint> geoPoints,
    double width = 3.0,
    Color color = Colors.green,
    bool isDotted = false,
  }) {
    final points =
        GeoSerie(type: GeoSerieType.line, name: "serie", geoPoints: geoPoints)
            .toLatLng();
    _linesState.addLine(
      name: name,
      line: Polyline(
        points: points,
        color: color,
        strokeWidth: width,
        isDotted: isDotted,
      ),
    );
  }

  /// Add a line on the map.
  void addPolyline({required String name, required Polyline polyline}) {
    _linesState.addLine(name: name, line: polyline);
  }

  /// Remove a line from the map
  void removeLine(String name) => _linesState.removeLine(name);

  /// Remove multiple lines from the map
  void removeLines(List<String> names) => _linesState.removeLines(names);

  /// Remove a polygon from the map
  void removePolygon(String name) => _polygonsState.removePolygon(name);

  /// Remove multiple polygons from the map
  void removePolygons(List<String> names) =>
      _polygonsState.removePolygons(names);

  /// Add a polygon on the map
  void addPolygon({
    required String name,
    required List<LatLng> points,
    Color color = Colors.lightBlue,
    double borderWidth = 0.0,
    Color borderColor = const Color(0xFFFFFF00),
  }) =>
      _polygonsState.addPolygon(
        name: name,
        points: points,
        color: color,
        borderWidth: borderWidth,
        borderColor: borderColor,
      );

  /// Switch to a tile layer
  void switchTileLayer(TileLayerType layer) =>
      _tileLayerState.switchTileLayer(layer);

  /// Display some geojson data on the map
  Future<void> fromGeoJson(
    String data, {
    bool verbose = false,
    Icon markerIcon = const Icon(Icons.location_on),
    bool noIsolate = kIsWeb,
  }) async {
    debugPrint("From geojson $data");

    final geojson = GeoJson();
    geojson.processedFeatures.listen((GeoJsonFeature feature) {
      switch (feature.type) {
        case GeoJsonFeatureType.point:
          final point = feature.geometry as GeoJsonPoint;
          final pointName = point.name;
          if (pointName != null) {
            addMarker(
              name: pointName,
              marker: Marker(
                point:
                    LatLng(point.geoPoint.latitude, point.geoPoint.longitude),
                builder: (BuildContext context) => markerIcon,
              ),
            );
          }
          break;
        case GeoJsonFeatureType.multipoint:
          final mp = feature.geometry as GeoJsonMultiPoint;
          final geoSerie = mp.geoSerie;
          if (geoSerie != null) {
            for (final geoPoint in geoSerie.geoPoints) {
              final pointName = geoPoint.name;
              if (pointName != null) {
                addMarker(
                  name: pointName,
                  marker: Marker(
                    point: LatLng(geoPoint.latitude, geoPoint.longitude),
                    builder: (BuildContext context) => markerIcon,
                  ),
                );
              }
            }
          }
          break;
        case GeoJsonFeatureType.line:
          final line = feature.geometry as GeoJsonLine;
          final lineName = line.name;
          final geoSerie = line.geoSerie;
          if (lineName != null && geoSerie != null) {
            addLine(name: lineName, points: geoSerie.toLatLng());
          }
          break;
        case GeoJsonFeatureType.multiline:
          final ml = feature.geometry as GeoJsonMultiLine;
          for (final line in ml.lines) {
            final lineName = line.name;
            final geoSerie = line.geoSerie;
            if (lineName != null && geoSerie != null) {
              addLine(name: lineName, points: geoSerie.toLatLng());
            }
          }
          break;
        case GeoJsonFeatureType.polygon:
          final poly = feature.geometry as GeoJsonPolygon;
          for (final geoSerie in poly.geoSeries) {
            addPolygon(name: geoSerie.name, points: geoSerie.toLatLng());
          }
          break;
        case GeoJsonFeatureType.multipolygon:
          final mp = feature.geometry as GeoJsonMultiPolygon;
          for (final poly in mp.polygons) {
            for (final geoSerie in poly.geoSeries) {
              addPolygon(name: geoSerie.name, points: geoSerie.toLatLng());
            }
          }
          break;
        case GeoJsonFeatureType.geometryCollection:
          // TODO : implement
          throw const NotImplementedException(
            "GeoJsonFeatureType.geometryCollection Not implemented",
          );
      }
    });
    if (noIsolate) {
      await geojson.parseInMainThread(data);
    } else {
      await geojson.parse(data);
    }
  }

  /// Export all the map assets to a [GeoJsonFeatureCollection]
  GeoJsonFeatureCollection toGeoJsonFeatures() {
    final featureCollection = GeoJsonFeatureCollection()
      ..collection = <GeoJsonFeature>[];
    final markersFeature = _markersState.toGeoJsonFeatures();
    final linesFeature = _linesState.toGeoJsonFeatures();
    final polygonsFeature = _polygonsState.toGeoJsonFeatures();
    if (markersFeature != null) {
      featureCollection.collection.add(markersFeature);
    }
    if (linesFeature != null) {
      featureCollection.collection.add(linesFeature);
    }
    if (polygonsFeature != null) {
      featureCollection.collection.add(polygonsFeature);
    }
    return featureCollection;
  }

  /// Convert the map assets to a geojson string
  String toGeoJson() => toGeoJsonFeatures().serialize();

  /// Notify to the changefeed
  void notify(
    String name,
    Object? value,
    Function from,
    MapControllerChangeType type,
  ) {
    final change = StatefulMapControllerStateChange(
      name: name,
      value: value,
      from: from,
      type: type,
    );
    if (verbose) {
      debugPrint("Map state change: $change");
    }
    _subject.add(change);
  }
}


example/lib/markers.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:latlong2/latlong.dart';
import 'package:map_controller_plus_3/map_controller_plus_3.dart';

class Place {
  Place(this.name, this.point);

  final String name;
  final LatLng point;
}

class _MarkersPageState extends State<MarkersPage> {
  final mapController = MapController();
  late final statefulMapController =
      StatefulMapController(mapController: mapController);
  late final StreamSubscription<StatefulMapControllerStateChange> sub;

  final List<Place> places = [
    Place("Notre-Dame", LatLng(48.853831, 2.348722)),
    Place("Montmartre", LatLng(48.886463, 2.341169)),
    Place("Champs-Elysées", LatLng(48.873932, 2.294821)),
    Place("Chinatown", LatLng(48.827393, 2.361897)),
    Place("Tour Eiffel", LatLng(48.85801, 2.294713))
  ];

  final _markersOnMap = <Place>[];
  bool ready = false;

  void addMarker(BuildContext context) {
    for (final place in places) {
      if (!_markersOnMap.contains(place)) {
        debugPrint("Adding marker ${place.name}");
        statefulMapController.addMarker(
            name: place.name,
            marker: Marker(
                point: place.point,
                builder: (BuildContext context) {
                  return const Icon(Icons.location_on);
                }));
        _markersOnMap.add(place);
        return;
      }
    }
  }

  @override
  void initState() {
    statefulMapController.onReady.then((_) => setState(() => ready = true));
    sub = statefulMapController.changeFeed.listen((change) => setState(() {}));
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: FlutterMap(
          mapController: mapController,
          options: MapOptions(
            center: LatLng(48.853831, 2.348722),
            zoom: 11.0,
          ),
          children: [
            statefulMapController.tileLayer!,
            MarkerLayer(
              markers: statefulMapController.markers,
            ),
          ],
        ),
      ),
      floatingActionButton: ready
          ? FloatingActionButton(
              child: const Icon(Icons.add),
              onPressed: () => addMarker(context),
            )
          : const Text(""),
    );
  }

  @override
  void dispose() {
    sub.cancel();
    super.dispose();
  }
}

class MarkersPage extends StatefulWidget {
  const MarkersPage({Key? key}) : super(key: key);

  @override
  _MarkersPageState createState() => _MarkersPageState();
}


lib/src/state/tile_layer.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';

import '../controller.dart';
import '../models.dart';
import '../types.dart';

/// The state of the tile layer
class TileLayerState {
  /// Default contructor
  TileLayerState({
    required this.type,
    required this.notify,
    this.customTileLayer,
  }) {
    _tileLayer = _tileLayerForType(type);
  }

  /// The tile layer type
  TileLayerType type;

  /// A custom tile layer options
  TileLayer? customTileLayer;

  /// Function to notify the changefeed
  final FeedNotifyFunction notify;

  TileLayer? _tileLayer;

  TileLayer? get tileLayer => _tileLayer;

  void switchTileLayer(TileLayerType layer) {
    _tileLayer = _tileLayerForType(layer);
    notify(
      "switchTileLayer",
      layer,
      switchTileLayer,
      MapControllerChangeType.tileLayer,
    );
  }

  TileLayer? _tileLayerForType(TileLayerType layer) {
    TileLayer? tlo;
    switch (layer) {
      case TileLayerType.hike:
        tlo = TileLayer(
          urlTemplate: "https://tiles.wmflabs.org/hikebike/{z}/{x}/{y}.png",
          subdomains: ['a', 'b', 'c'],
          tileProvider:
              kIsWeb ? _NonCachingNetworkTileProvider() : NetworkTileProvider(),
        );
        break;
      case TileLayerType.topography:
        tlo = TileLayer(
          urlTemplate: "http://{s}.tile.opentopomap.org/{z}/{x}/{y}.png",
          subdomains: ['a', 'b', 'c'],
          tileProvider:
              kIsWeb ? _NonCachingNetworkTileProvider() : NetworkTileProvider(),
        );
        break;
      case TileLayerType.monochrome:
        tlo = TileLayer(
          urlTemplate:
              "http://www.toolserver.org/tiles/bw-mapnik/{z}/{x}/{y}.png",
          subdomains: ['a', 'b', 'c'],
          tileProvider:
              kIsWeb ? _NonCachingNetworkTileProvider() : NetworkTileProvider(),
        );
        break;
      case TileLayerType.custom:
        if (customTileLayer != null) {
          tlo = customTileLayer;
        } else {
          tlo = customTileLayer;
        }
        break;
      default:
        tlo = TileLayer(
          urlTemplate: "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
          subdomains: ['a', 'b', 'c'],
          tileProvider:
              kIsWeb ? _NonCachingNetworkTileProvider() : NetworkTileProvider(),
        );
    }
    return tlo;
  }
}

class _NonCachingNetworkTileProvider extends TileProvider {
  @override
  ImageProvider getImage(Coords<num> coords, TileLayer options) {
    return NetworkImage(getTileUrl(coords, options));
  }
}


lib/src/tile_layer/tile_layers_bar.dart
import 'dart:async';

import 'package:flutter/material.dart';

import '../../map_controller_plus_3.dart';

class _TileLayersBarState extends State<TileLayersBar> {
  late final StatefulMapController controller = widget.controller;
  late TileLayerType _tileLayerType = controller.tileLayerType;
  late StreamSubscription<StatefulMapControllerStateChange> _sub;

  @override
  void initState() {
    _sub = controller.changeFeed.listen((change) {
      if (change.type == MapControllerChangeType.tileLayer) {
        setState(() => _tileLayerType = change.value as TileLayerType);
      }
    });
    super.initState();
  }

  Widget _buildLayers() {
    return Column(
      children: <Widget>[
        IconButton(
          iconSize: 30.0,
          color: (_tileLayerType == TileLayerType.normal)
              ? Colors.blueGrey
              : Colors.grey,
          icon: const Icon(Icons.map),
          tooltip: "Normal layer",
          onPressed: () => controller.switchTileLayer(TileLayerType.normal),
        ),
        IconButton(
          iconSize: 30.0,
          color: (_tileLayerType == TileLayerType.monochrome)
              ? Colors.blueGrey
              : Colors.grey,
          icon: const Icon(Icons.local_car_wash),
          tooltip: "Monochrome layer",
          onPressed: () => controller.switchTileLayer(TileLayerType.monochrome),
        ),
        IconButton(
          iconSize: 30.0,
          color: (_tileLayerType == TileLayerType.topography)
              ? Colors.blueGrey
              : Colors.grey,
          icon: const Icon(Icons.photo),
          tooltip: "Topography layer",
          onPressed: () => controller.switchTileLayer(TileLayerType.topography),
        ),
        IconButton(
          iconSize: 30.0,
          color: (_tileLayerType == TileLayerType.hike)
              ? Colors.blueGrey
              : Colors.grey,
          icon: const Icon(Icons.landscape),
          tooltip: "Hills layer",
          onPressed: () => controller.switchTileLayer(TileLayerType.hike),
        ),
      ],
    );
  }

  @override
  Widget build(BuildContext context) => _buildLayers();

  @override
  void dispose() {
    _sub.cancel();
    super.dispose();
  }
}

/// The tile layers bar
class TileLayersBar extends StatefulWidget {
  /// Provide a controller
  const TileLayersBar({Key? key, required this.controller}) : super(key: key);

  /// The map controller
  final StatefulMapController controller;

  @override
  State<TileLayersBar> createState() => _TileLayersBarState();
}


example/lib/stateful_markers.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:flutter_map/flutter_map.dart';
import 'package:latlong2/latlong.dart';
import 'package:map_controller_plus_3/map_controller_plus_3.dart';

class Place {
  Place(this.name, this.point);

  final String name;
  final LatLng point;
}

class _StatefulMarkersPageState extends State<StatefulMarkersPage> {
  final mapController = MapController();
  late final statefulMapController =
      StatefulMapController(mapController: mapController);
  late final StreamSubscription<StatefulMapControllerStateChange> sub;

  final List<Place> places = [
    Place("Notre-Dame", LatLng(48.853831, 2.348722)),
    Place("Montmartre", LatLng(48.886463, 2.341169)),
    Place("Champs-Elysées", LatLng(48.873932, 2.294821)),
    Place("Chinatown", LatLng(48.827393, 2.361897)),
    Place("Tour Eiffel", LatLng(48.85801, 2.294713))
  ];

  final _markersOnMap = <Place>[];
  bool ready = false;

  void addMarker(BuildContext context) {
    for (final place in places) {
      if (!_markersOnMap.contains(place)) {
        debugPrint("Adding marker ${place.name}");
        statefulMapController.addStatefulMarker(
            name: place.name,
            statefulMarker: StatefulMarker(
                //anchorAlign: AnchorAlign.bottom,
                height: 80.0,
                width: 150.0,
                state: <String, dynamic>{"showText": false},
                point: place.point,
                builder: (context, state) {
                  Widget w;
                  final markerIcon = IconButton(
                      icon: const Icon(Icons.location_on),
                      onPressed: () => statefulMapController.mutateMarker(
                          name: place.name,
                          property: "showText",
                          value: !(state["showText"] as bool)));
                  if (state["showText"] == true) {
                    w = Column(children: <Widget>[
                      markerIcon,
                      Container(
                          color: Colors.white,
                          child: Padding(
                              padding: const EdgeInsets.all(5.0),
                              child: Text(place.name, textScaleFactor: 1.3))),
                    ]);
                  } else {
                    w = markerIcon;
                  }
                  return w;
                }));
        _markersOnMap.add(place);
        return;
      }
    }
  }

  @override
  void initState() {
    statefulMapController.onReady.then((_) => setState(() => ready = true));
    sub = statefulMapController.changeFeed.listen((change) => setState(() {}));
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: FlutterMap(
          mapController: mapController,
          options: MapOptions(
            center: LatLng(48.853831, 2.348722),
            zoom: 11.0,
          ),
          children: [
            statefulMapController.tileLayer!,
            MarkerLayer(
                markers: statefulMapController
                    .getMarkers(['Notre-Dame', 'Montmartre'])),
          ],
        ),
      ),
      floatingActionButton: ready
          ? FloatingActionButton(
              child: const Icon(Icons.add),
              onPressed: () => addMarker(context),
            )
          : const Text(""),
    );
  }

  @override
  void dispose() {
    sub.cancel();
    super.dispose();
  }
}

class StatefulMarkersPage extends StatefulWidget {
  const StatefulMarkersPage({Key? key}) : super(key: key);

  @override
  _StatefulMarkersPageState createState() => _StatefulMarkersPageState();
}
